src/plots/Mesh/avtMeshPlot.C

```diff
@@ -347,21 +349,71 @@ avtMeshPlot::SetAtts(const AttributeGroup *a)
     atts = *(const MeshAttributes*)a;
 
     SetLineWidth(Int2LineWidth(atts.GetLineWidth()));
-    if (atts.GetMeshColorSource()==0)
+    if (atts.GetMeshColorSource() == MeshAttributes::Foreground)
     {
         SetMeshColor(fgColor);
     }
-    else
+    else if (atts.GetMeshColorSource() == MeshAttributes::MeshRandom)
+    {
+        unsigned char rgb[3] = {0,0,0};
+        unsigned char fg[3] = {static_cast<unsigned char>(fgColor[0]*255),
+                               static_cast<unsigned char>(fgColor[1]*255),
+                               static_cast<unsigned char>(fgColor[2]*255)};
+
+        //
+        // Try and get the color for the colorIndex'th color in the distinct
+        // discrete color table or, failing that, default discrete color table.
+        // NOTE: The index is mod'd by the number of control points.
+        //
+        avtColorTables *ct = avtColorTables::Instance();
+        int colorIndex = this->instanceIndex;
+        do
+        {
+            if (! ct->GetControlPointColor(ct->GetDefaultDiscreteColorTable(), colorIndex, rgb))
+                if (! ct->GetControlPointColor("distinct", colorIndex, rgb))
+                    break;
+            colorIndex++;
+        }
+        while ((colorIndex - this->instanceIndex) < 50 &&
+               (ct->PerceptualColorDistance(rgb, fg) <= ct->JustNoticeableColorDistance()));
+        SetMeshColor(rgb);
+    }
+    else // MeshAttributes::MeshCustom
     {
         SetMeshColor(atts.GetMeshColor().GetColor());
     }
 
     SetRenderOpaque();
-    if (atts.GetOpaqueColorSource()==0)
+    if (atts.GetOpaqueColorSource() == MeshAttributes::Background)
     {
         SetOpaqueColor(bgColor);
     }
-    else
+    else if (atts.GetOpaqueColorSource() == MeshAttributes::OpaqueRandom)
+    {
+        unsigned char rgb[3] = {0,0,0};
+        unsigned char bg[3] = {static_cast<unsigned char>(bgColor[0]*255),
+                               static_cast<unsigned char>(bgColor[1]*255),
+                               static_cast<unsigned char>(bgColor[2]*255)};
+
+        //
+        // Try and get the color for the colorIndex'th color in the distinct
+        // discrete color table or, failing that, default discrete color table.
+        // NOTE: The index is mod'd by the number of control points.
+        //
+        avtColorTables *ct = avtColorTables::Instance();
+        int colorIndex = this->instanceIndex + 1;
+        do
+        {
+            if (! ct->GetControlPointColor(ct->GetDefaultDiscreteColorTable(), colorIndex, rgb))
+                if (! ct->GetControlPointColor("distinct", colorIndex, rgb))
+                    break;
+            colorIndex++;
+        }
+        while ((colorIndex - this->instanceIndex) < 50 &&
+               (ct->PerceptualColorDistance(rgb, bg) <= ct->JustNoticeableColorDistance()));
```

Comment:
While very cool, does it really matter if the 'opaque' surface randomly becomes the same color as the background? That is the default mode and perhaps a valid color even in random mode?

Reply:
Maybe it doesn't matter. I does matter if the opaque color randomly becomes the same color as the mesh color. However, the only way I have to try to avoid that was to start at a different index in the colortable, hence the `+1` in initializing `colorIndex`. I don't have a way to ask the mesh what is current color is.